File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
    private:
        bool foo(vector<int> candidates, int target, int idx, int sum, vector<int>& cur, vector<vector<int>>& res){
            
            int n=candidates.size();
            
            if(idx>candidates.size() || sum>target){
                return false;
            }
            
            if(sum==target){
                res.push_back(cur);
            }
            
            
            for(int i=idx;i<n;i++){
                int ele=candidates[i];
                cur.push_back(ele);
                foo(candidates, target, i, sum+ele, cur, res);
                cur.pop_back();
            }
            
            return true;
            
        }
    public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> cur;
        vector<vector<int>> res;
        sort(candidates.begin(), candidates.end());
        
        foo(candidates, target, 0, 0, cur, res);
        return res;
        
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        cur=[]
        ans=[]
        sum=0
        n=len(candidates)
        def dfs(idx, cur, ans, sum):
            if(sum==target):
                ans.append(cur.copy())
                return 
            if(sum>=target):
                return;
        
            for i in range(idx,n):
                ele=candidates[i]
                cur.append(ele)
                dfs(i, cur, ans, sum+ele)
                cur.pop()
        dfs(0,cur,ans, sum)
        return ans